### PR TITLE
[IMP] mail, *: move channel concerns in Thread from Thread to Channel

### PR DESCRIPTION
**Pull request by oomsveta**
This commit is the continuation of task-2948676 which consists of lightening the Thread model by moving the channel-specific behaviours to the Channel model.

The changes revolve around the DiscussSidebarCategoryItem model. The goal is to use channel instead of thread as identifying field for DiscussSidebarCategoryItem, and to move channel-specific fields involved in the computes of this model from Thread to Channel.

Changes in this commit include:
+ move avatarCacheKey from Thread to Channel
+ move correspondent
+ move correspondentOfDmChat
+ move custom_channel_name
+ move localMessageUnreadCounter
+ move serverMessageUnreadCounter
+ introduce Channel/displayName
+ use channel instead of thread as identifying field for DiscussSidebarCategoryItem

Part of task-2948676.
Enterprise: odoo/enterprise#30457
\* = hr_holidays, im_livechat, test_discuss_full

### CODE REVIEW

File: addons/mail/static/src/models/channel.js [hunk 1]
```diff
@@ -60,6 +60,20 @@ registerModel({
...
         },
+        /**
+         * @private
+         * @returns {Partner|FieldCommand}
+         */
+        _computeCorrespondentOfDmChat() {
+            if (
+                this.channel_type === 'chat' &&
+                this.correspondent &&
+                this.thread.public === 'private'
```
Comment (seb-odoo) [inline addons/mail/static/src/models/channel.js:71]: this condition is useless, and will be removed soon in another PR

File: addons/mail/static/src/models/channel.js [hunk 2]
```diff
@@ -136,13 +151,19 @@ registerModel({
...
         id: attr({
             identifying: true,
         }),
         /**
          * States the number of members in this channel according to the server.
          */
         memberCount: attr(),
+        name: attr({
+            related: 'thread.name',
+        }),
```
Comment (seb-odoo) [inline addons/mail/static/src/models/channel.js:256]: I don't like this kind of related, I'd rather see `thread.name` directly in code.

File: addons/mail/static/src/models/channel.js [hunk 3]
```diff
@@ -185,6 +185,10 @@ registerModel({
             compute: '_computeCorrespondentOfDmChat',
             inverse: 'dmChatWithCurrentPartner',
         }),
+        /**
+         * FIXME: should rather be on ChannelMember.
+         */
```
Comment (alexkuhn) [inline addons/mail/static/src/models/channel.js:226]: Remove FIXME, the modelling will be changed when needed, and it's more appropriate to actually rename field and/or move it to another model when it feels best.

**Review by alexkuhn — COMMENTED** [review]
LGTM otherwise

File: addons/mail/static/src/models/channel.js (hunk 1 above)
Comment (oomsveta) [reply addons/mail/static/src/models/channel.js:71]: the check on `public === 'private'`? why? are chat channels always private?

File: addons/mail/static/src/models/thread_preview_view.js [hunk 4]
```diff
@@ -75,10 +75,16 @@ registerModel({
...
          */
         _computePersonaImStatusIconView() {
-            return this.thread.correspondent && this.thread.correspondent.isImStatusSet ? {} : clear();
+            if (!this.thread.channel.correspondent) {
+                return clear();
+            }
+            if (this.thread.channel.correspondent.isImStatusSet) {
+                return {};
+            }
+            return clear();
```
[the remark this answers is not in the document]
Comment (oomsveta) [reply addons/mail/static/src/models/thread_preview_view.js:87]: I prefer this syntax because it highlights what is merely a guard and what is the actual condition.

File: addons/mail/static/src/components/discuss_sidebar_category_item/discuss_sidebar_category_item.xml [hunk 5]
```diff
@@ -8,8 +8,8 @@
                     'bg-100': !categoryItem.isActive,
                     'o-active bg-200': categoryItem.isActive,
                     'o-unread': categoryItem.isUnread,
-                }" t-attf-class="{{ className }}" t-on-click="categoryItem.onClick" t-att-data-thread-id="categoryItem.thread.id" t-att-data-thread-name="categoryItem.thread.displayName"
-                t-att-data-thread-model="categoryItem.thread.model" t-ref="root"
+                }" t-attf-class="{{ className }}" t-on-click="categoryItem.onClick" t-att-data-channel-id="categoryItem.channel.id" t-att-data-channel-name="categoryItem.channel.displayName"
```
[the remark this answers is not in the document]
Comment (oomsveta) [reply addons/mail/static/src/components/discuss_sidebar_category_item/discuss_sidebar_category_item.xml:11]: I can find some occurrences in tests

File: addons/mail/models/mail_channel.py [hunk 6]
```diff
@@ -1021,8 +1021,11 @@ def channel_set_custom_name(self, name):
         member = self.env['mail.channel.member'].search([('partner_id', '=', self.env.user.partner_id.id), ('channel_id', '=', self.id)])
         member.write({'custom_channel_name': name})
         self.env['bus.bus']._sendone(member.partner_id, 'mail.channel/insert', {
+            'channel': {
+                'custom_channel_name': name,
+                'id': self.id,
+            },
```
[the remark this answers is not in the document]
Comment (oomsveta) [reply addons/mail/models/mail_channel.py:1028]: great idea

File: addons/mail/static/src/components/thread_preview/thread_preview.js [hunk 7]
```diff
@@ -25,10 +25,13 @@ export class ThreadPreview extends Component {
      * @returns {string}
      */
     image() {
+        if (!this.threadPreviewView.thread.channel) {
+            return '/mail/static/src/img/smiley/avatar.jpg';
+        }
```
[the remark this answers is not in the document]
Comment (oomsveta) [reply addons/mail/static/src/components/thread_preview/thread_preview.js]: As `correspondent` was directly on Thread there was no need for it to be guarded.
I agree with you in that I don't think there are any cases where the component doesn't have a channel. Actually, ThreadPreview should rather be ChannelPreview, but I don't intend to make the change in this PR, so as long as we have to deal with thread instead of channel directly I'd rather keep some kind of guard just in case.
Be sure however that it will be addressed in a further PR 🙂

File: addons/mail/static/src/models/channel.js [hunk 8]
```diff
@@ -140,6 +160,22 @@ registerModel({
...
+        _getDiscussSidebarCategory() {
+            switch (this.channel_type) {
+                case 'channel':
+                    return this.messaging.discuss.categoryChannel;
+                case 'chat':
+                case 'group':
+                    return this.messaging.discuss.categoryChat;
+            }
+        },
```
[the remark this answers is not in the document]
Comment (oomsveta) [reply addons/mail/static/src/models/channel.js]: done

**Review by alexkuhn — APPROVED** [review]

File: addons/mail/static/src/models/thread_preview_view.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/models/thread_preview_view.js:87]: Both are actual conditions, the existence of the correspondent is not a technical detail, it's an important part of determining if there is im status.

**Comment by alexkuhn** [comment]
@robodoo r+
